### PR TITLE
Allow make_xarray_grid to take horizontal coordinates as 1d arrays

### PR DESCRIPTION
**Pull request by santisoler**
Now `make_xarray_grid` could take 1d arrays for horizontal coordinates, such as
how `xr.DataArray` stores the single dimension coordinates. Add a new
non-public `meshgrid_to_1d` function to convert 2d meshgrids (like the ones
generated by `vd.grid_coordinates`) to 1d arrays containing a single coordinate
of the corresponding axis. This function checks if all elements inside the 2d
array are equal along the right axis, i.e. a vaild meshgrid.

**Reminders**:

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst` and the base `__init__.py` file for the package.
- [ ] Write detailed docstrings for all functions/classes/methods. It often helps to design better code if you write the docstrings first.
- [ ] If adding new functionality, add an example to the docstring, gallery, and/or tutorials.
- [ ] Add your full name, affiliation, and [ORCID](https://orcid.org) (optional) to the `AUTHORS.md` file (if you haven't already) in case you'd like to be listed as an author on the [Zenodo](https://zenodo.org/communities/fatiando) archive of the next release.

### CODE REVIEW

File: doc/api/index.rst [hunk 1]
```diff
@@ -93,6 +93,7 @@ Utilities
     variance_to_weights
     grid_to_table
     make_xarray_grid
+    meshgrid_to_1d
```
Comment (santisoler) [inline doc/api/index.rst:96]:
```suggestion
```

File: verde/__init__.py [hunk 2]
```diff
@@ -15,7 +15,13 @@
     longitude_continuity,
 )
 from .mask import distance_mask, convexhull_mask
-from .utils import variance_to_weights, maxabs, grid_to_table, make_xarray_grid
+from .utils import (
+    variance_to_weights,
+    maxabs,
+    grid_to_table,
+    make_xarray_grid,
+    meshgrid_to_1d,
```
Comment (santisoler) [inline verde/__init__.py:23]:
```suggestion
```

File: verde/utils.py [hunk 3]
```diff
@@ -327,6 +338,82 @@ def make_xarray_grid(
...
+    check_coordinates(coordinates)
+    _check_meshgrid(coordinates)
+    easting, northing = coordinates[0][0, :], coordinates[1][:, 0]
+    coordinates = (easting, northing, *coordinates[2:])
+    return coordinates
+
+
+def _check_meshgrid(coordinates):
+    """
+    Check if the horizontal coordiantes of a regular grid are meshgrids
```
Comment (santisoler) [inline verde/utils.py:391]:
```suggestion
    Check if the horizontal coordinates of a regular grid are meshgrids
```

File: verde/tests/test_utils.py [hunk 4]
```diff
@@ -212,3 +213,78 @@ def test_make_xarray_grid_invalid_extra_coords():
         make_xarray_grid(
             coordinates, data, data_names="dummy", extra_coords_names=["upward", "time"]
         )
+
+
+def test_make_xarray_grid_invalid_2d_coordiantes():
```
Comment (santisoler) [inline verde/tests/test_utils.py:218]:
```suggestion
def test_make_xarray_grid_invalid_2d_coordinates():
```

File: verde/tests/test_utils.py [hunk 5]
```diff
@@ -212,3 +213,78 @@ def test_make_xarray_grid_invalid_extra_coords():
...
+    """
+    Check if error is raised if invaild 2d coordiantes array are passed
```
Comment (santisoler) [inline verde/tests/test_utils.py:220]:
```suggestion
    Check if error is raised if invaild 2d coordinates array are passed
```

File: verde/tests/test_utils.py [hunk 6]
```diff
@@ -212,3 +213,78 @@ def test_make_xarray_grid_invalid_extra_coords():
...
+    # easting is 1d, but northing is 2d
+    easting = easting[0, :]
+    data = np.ones_like(easting)
+    with pytest.raises(ValueError):
+        make_xarray_grid((easting, northing), data, data_names="dummy")
+    # northing is 1d, but easting is 2d
+    northing = northing[:, 0]
+    data = np.ones_like(easting)
+    with pytest.raises(ValueError):
+        make_xarray_grid((easting, northing), data, data_names="dummy")
```
Comment (leouieda) [inline verde/tests/test_utils.py:263]: The way this is now, easting is 1d for the second check. 

```suggestion
    data = np.ones_like(easting)
    # easting is 1d, but northing is 2d
    with pytest.raises(ValueError):
        make_xarray_grid((easting[0, :], northing), data, data_names="dummy")
    # northing is 1d, but easting is 2d
    with pytest.raises(ValueError):
        make_xarray_grid((easting, northing[:, 0]), data, data_names="dummy")
```

File: verde/utils.py [hunk 7]
```diff
@@ -311,12 +311,23 @@ def make_xarray_grid(
         dummy     (northing, easting) float64 1.0 1.0 1.0 1.0 1.0 1.0
 
     """
-    coordinates = check_coordinates(coordinates)
+    # Check dimensions of the horizontal coordinates of the regular grid
+    ndim = np.ndim(coordinates[0])
+    if ndim != np.ndim(coordinates[1]):
+        raise ValueError(
+            "Incompatible dimensions between horizontal coordinates of a regular grid: "
```
Comment (leouieda) [inline verde/utils.py:318]:
```suggestion
            "Incompatible dimensions between horizontal coordinates of the regular grid: "
```

File: verde/utils.py [hunk 8]
```diff
@@ -311,12 +311,23 @@ def make_xarray_grid(
...
+            + f"'{ndim}' and '{np.ndim(coordinates[1])}'. "
+            + "Horizontal coordinates of a regular grid must have the same "
+            + "number of dimensions"
```
Comment (leouieda) [inline verde/utils.py:321]:
```suggestion
            + "Both coordinates must have the same number of dimensions."
```

File: verde/utils.py [hunk 9]
```diff
@@ -327,6 +338,82 @@ def make_xarray_grid(
...
+    [[2. 2. 2. 2. 2.]
+     [2. 2. 2. 2. 2.]
+     [2. 2. 2. 2. 2.]
+     [2. 2. 2. 2. 2.]
+     [2. 2. 2. 2. 2.]
+     [2. 2. 2. 2. 2.]
+     [2. 2. 2. 2. 2.]]
+    """
+    check_coordinates(coordinates)
+    _check_meshgrid(coordinates)
```
Comment (leouieda) [inline verde/utils.py:383]:
```suggestion
    check_regular_grid(coordinates)
```

File: verde/utils.py [hunk 10]
```diff
@@ -327,6 +338,82 @@ def make_xarray_grid(
...
+    easting, northing = coordinates[0][0, :], coordinates[1][:, 0]
+    coordinates = (easting, northing, *coordinates[2:])
+    return coordinates
+
+
+def _check_meshgrid(coordinates):
```
Comment (leouieda) [inline verde/utils.py:389]:
```suggestion
def check_regular_grid(coordinates):
```

File: verde/utils.py [hunk 11]
```diff
@@ -327,6 +338,82 @@ def make_xarray_grid(
...
+    """
+    Check if the horizontal coordinates of a regular grid are meshgrids
```
Comment (leouieda) [inline verde/utils.py:391]:
```suggestion
    Check if the horizontal coordinates given define a regular grid
    
    Does not check if the grid is evenly spaced. Instead, check if
    the easting coordinates don't vary with rows and that the 
    northing don't vary with columns of the respective arrays. 
```

**Review by leouieda — APPROVED** [review]
Looks really good @santisoler! Just a few minor suggestions and feel free to merge afterwards.

File: verde/utils.py [hunk 12]
```diff
@@ -333,6 +343,86 @@ def make_xarray_grid(
...
+    check_coordinates(coordinates)
+    check_regular_grid(coordinates)
+    easting, northing = coordinates[0][0, :], coordinates[1][:, 0]
+    coordinates = (easting, northing, *coordinates[2:])
+    return coordinates
+
+
+def check_regular_grid(coordinates):
+    """
+    Check if the horizontal coordinates given define a regular grid
```
Comment (santisoler) [inline verde/utils.py:396]:
```suggestion
    Check if the given horizontal coordinates define a regular grid
```

File: verde/utils.py [hunk 13]
```diff
@@ -333,6 +343,86 @@ def make_xarray_grid(
...
+    easting, northing = coordinates[0][0, :], coordinates[1][:, 0]
+    coordinates = (easting, northing, *coordinates[2:])
+    return coordinates
+
+
+def check_regular_grid(coordinates):
+    """
+    Check if the given horizontal coordinates define a regular grid
+
+    Does not check if the grid is evenly spaced. Instead, check if
```
Comment (santisoler) [inline verde/utils.py:398]:
```suggestion
    Don't not check if the grid is evenly spaced. Instead, check if
```

**Comment by santisoler** [comment]
@leouieda Since we renamed the check function to `check_regular_grid` I added an additional check: if the horizontal coordinates of a regular grid are evenly spaced.

**Comment by leouieda** [comment]
@santisoler i don’t think that check is necessary since we can still do the conversion to xarray if the spacing is variable. So maybe that function could really be `check_meshgrid` since it just needs to verify that the 2D arrays come from a meshgrid call. Sorry about that 🙂

**Comment by santisoler** [comment]
> @santisoler i don’t think that check is necessary since we can still do the conversion to xarray if the spacing is variable. So maybe that function could really be `check_meshgrid` since it just needs to verify that the 2D arrays come from a meshgrid call. Sorry about that slightly_smiling_face

No problem! I though the same when I named the function. When you suggested to change it to `check_regular_grid` I though the function should also check if the coordinates are actually evenly spaced. But you're right, for the purpose of the function it's ok to just check if the 2d-arrays are returns from a meshgrid call.

Let me revert the changes and rename the function to `check_meshgrid`.

**Review by leouieda — APPROVED** [review]

**Comment by leouieda** [comment]
👍🏽 this seems good to go!

**Pull request merged**